Multicontainer updates, round one

zwalchuk opened the pull request:
A big batch of docs updates to reflect multicontainer changes, including a multicontainer guide, updates to service and configuration variable pages, changes to getting started guides, and a whole bunch of new screenshots.

Fixes #201, #338, #500

LucianBuzzo commented on pages/deployment/multicontainer.md:
```diff
@@ -0,0 +1,80 @@
+
+# Multicontainer Guide
+
+As your applications grow more complex, you may find significant benefit in running some services in separate containers. Multicontainer applications allow you to better isolate, duplicate, and manage key services, making your applications more modular and secure. Starting with resinOS vx.x.x, resin.io provides multicontainer application support, built around the [Docker Compose][docker-compose]  ...
```
Do we not know which version of resin OS will support multicontainer yet?

LucianBuzzo commented on pages/deployment/multicontainer.md:
````diff
@@ -0,0 +1,80 @@
+    build: ./redis
+```
+
+Each service can either be built from a directory containing a Dockerfile, as shown here, or can use a **Docker** image that has already been built, by replacing `build:` with `image:`. If your containers need to started in a specific order, make sure to use the `depends_on:` setting.
````
It would be nice to link to the `depends_on` docs https://docs.docker.com/compose/compose-file/compose-file-v2/#depends_on

LucianBuzzo commented on pages/deployment/multicontainer.md:
```diff
@@ -0,0 +1,80 @@
+
+## Docker Compose
+
+At the root of your multicontainer application, you'll use a `docker-compose.yml` file to specify the configuration of your containers. The `docker-compose.yml` defines the services you'll be building, as well as how the services interact with each other and the host OS.
```
Can we briefly explain what docker compose is, what version(s) we support and link to the main docker compose documentation?

LucianBuzzo commented on pages/deployment/multicontainer.md:
```diff
@@ -0,0 +1,80 @@
+
+## Single container applications 
+
+If you push a project with only a Dockerfile or Dockerfile.template, rather than a `docker-compose.yml`, a single container will be built and downloaded, and the experience will be very similar to using resin.io before multicontainer support. In this form the single container, now known as a service, will be automatically assigned the name `main`, and you will see this service in a list on the de ...
```
The line containing "before multicontainer support" probably needs to be reworked. If I'm a new user, this information is meaningless. I think a more accurate description here would be 
> ...and the experience will be very similar to using resin.io with multiple containers.
The note below should be enough to help existing users who are curious about how their apps will now behave.

shaunmulligan replied:
We should perhaps show what the auto-generated docker-compose.yml for that single service would look like, just so people have an idea

shaunmulligan commented on pages/deployment/multicontainer.md:
```diff
@@ -0,0 +1,80 @@
+
+As your applications grow more complex, you may find significant benefit in running some services in separate containers. Multicontainer applications allow you to better isolate, duplicate, and manage key services, making your applications more modular and secure. Starting with resinOS vx.x.x, resin.io provides multicontainer application support, built around the [Docker Compose][docker-compose]  ...
+
+## Docker Compose
```
we should link to a full reference of all the things we currently support. I know @pcarranzav has a compiled list in a spread sheet

shaunmulligan commented on pages/deployment/multicontainer.md:
```diff
@@ -0,0 +1,80 @@
+Label | Default | Description
+--- | --- | ---
+io.resin.features.dbus | false | Bind mounts the hostOS dbus into the container using “/run/dbus:/host/run/dbus”
+io.resin.features.kernel_modules | false | Bind mounts the hostOS /lib/modules into the container. (i.e. “/lib/modules:/lib/modules”)
```
I believe all the feature labels here should have `-` rather than `_` in them, so `io.resin.features.kernel_modules` --> `io.resin.features.kernel-modules` as `_` are unsupported by docker

zwalchuk commented on pages/deployment/deployment.md:
```diff
@@ -40,27 +40,25 @@ specified, it defaults to running `node server.js`.)
 #### Node.js Example
 
 A good example of this is the [text-to-speech][text-to-speech] application -
```
@shaunmulligan the nice thing about the text-to-speech app is it shows how to use a preinstall script for installing dependencies with apt-get, which parallels nicely with the Dockerfiles.

zwalchuk commented on pages/deployment/docker-compose.md:
```diff
@@ -0,0 +1,72 @@
+
+# Docker Compose
+
+## Supported Compose fields
```
@pcarranzav I pulled these from the multicontainer preview guide, I'll need your guidance in figuring out how much of this is still valid.

CameronDiver commented on pages/deployment/multicontainer.md:
```diff
@@ -0,0 +1,131 @@
+
+## Docker Compose
+
+The multicontainer functionality provided by resin.io is built around **[Docker Compose][docker-compose]**, an open-source tool for running multiple **Docker** containers and coordinating resources between services. The resin.io device supervisor implements a subset of the [Compose v2.1 feature set][compose-features]. You can find a full list of supported and known unsupported features in our [de ...
```
This is kind of misleading, as I would assume from that that we're using docker's open source tool, but we're just using their file format.

CameronDiver commented on pages/deployment/multicontainer.md:
````diff
@@ -0,0 +1,131 @@
+restart: always
+```
+
+Setting `network_mode` to `host` allows the to container to share the same network namespace as the host OS. When this is set, any ports exposed on the container will be exposed locally on the device. This is necessary for features such as bluetooth.
````
`allows the to container`

CameronDiver commented on pages/management/configuration.md:
```diff
@@ -0,0 +1,81 @@
+
+Configuration variables allow you to provide runtime configuration to the host OS and supervisor. These variables all begin with `RESIN_`. Beginning with supervisor v7.0.0, a number of them appear automatically in your dashboard when your device is provisioned.
+
+Configuration variables can be managed at the fleet and device levels.
```
I think this would read better as 'Configuration variables can be managed at both the fleet and device level'

CameronDiver commented on pages/management/configuration.md:
```diff
@@ -0,0 +1,81 @@
+
+## Variable List
+
+This list contains configuration variables that can be used with resin.io devices, some of which will automatically appear for devices with supervisor v7.0.0 and greater. While they may not automatically populate in the dashboard, most of these variables can still be used with older supervisor versions, so be sure to check the *Valid for* column:
```
`Valid for` -> `Valid from`

CameronDiver commented on pages/management/configuration.md:
```diff
@@ -0,0 +1,81 @@
+
+<img src="/img/configuration/activate_default_config.png" width="100%">
+
+A new window will pop up, allowing you to define the value for your configuration variable. Click *Add* to apply to all devices in your fleet that do not already have an identical device configuration variable defined:
```
`window` -> `dialog`

CameronDiver commented on pages/management/configuration.md:
```diff
@@ -0,0 +1,81 @@
+
+<img src="/img/configuration/add_fleet_configuration.png" width="100%">
+
+As before, a window pops up giving you the option to define a name and value (remember, all config variable names must begin with `RESIN_`). Click *Add* to apply:
```
`window` -> `dialog`

CameronDiver commented on pages/management/configuration.md:
```diff
@@ -0,0 +1,81 @@
+
+## Managing Device Configuration Variables
+
+Device configuration variables define the behavior of one device. If both the fleet and the device have a configuration variable of the same name, the code on the device will use the value of the device configuration variable. In other words, device configuration variables redefine (or override) fleet configuration variables.
```
`of one device` -> `of a single device`

CameronDiver commented on pages/management/configuration.md:
```diff
@@ -0,0 +1,81 @@
+
+<img src="/img/configuration/device_config_variables.png" width="100%">
+
+This will pop up a small window for editing the value:
```
`window` -> `dialog`

CameronDiver commented on pages/management/configuration.md:
```diff
@@ -0,0 +1,81 @@
+
+Adding a custom device configuration variable is very similar to adding a custom fleet configuration variable: scroll past the prepopulated variables, click *Add custom variable*, define a name and value, and click *Add*.
+
+You can override the value of a custom fleet configuration variable by clicking *override* in the far-right column. This will pop up the variable editing window, where you can change the value:
```
`window` -> `dialog`

CameronDiver commented on pages/management/env-vars.md:
```diff
@@ -1,74 +1,56 @@
+# Service Variables 
 
-Environment variables allow you to provide runtime configuration to your application which can be updated without having to modify your source code.
+Service variables allow you to provide runtime configuration to a running service without having to modify your source code. You can use them to keep secrets and other sensitive values out of your code base.
```
`code base` -> `codebase`

CameronDiver commented on pages/management/env-vars.md:
```diff
@@ -1,74 +1,56 @@
+__Note:__ Adding or modifying a service variable will cause the service to restart on any devices where the new setting is applied.
 
-__Note:__ Adding or modifying a environment will cause the application to restart on the affected device/s.
+Resin.io allows you to set service variables at two levels: fleet and device.
```
I think this would be better as `to set service variables at both the device and fleet level.

CameronDiver commented on pages/management/env-vars.md:
```diff
@@ -1,74 +1,56 @@
+## Fleet Service Variables
 
-## Application-wide
+Service variables defined on the fleet level are available to the specified service on any device in that fleet, unless they are redefined with a device service variable of the same name.
```
`on any device` -> `on all devices`

CameronDiver commented on pages/management/env-vars.md:
```diff
@@ -1,74 +1,56 @@
+<img src="/img/env-vars/add_application_variable.png" width="40%">
 
-It's important to understand that the device environment variable _doesn't have to_ correspond to some application-wide environment variable.
+A new window will pop up, asking you to select a service from the drop down menu. Once you have selected the service, you can define the name and value for your service variable. Click *Add* to apply to all devices in your fleet that do not already have an identical device service variable defined:
```
`window` -> `dialog`

CameronDiver commented on pages/management/env-vars.md:
```diff
@@ -1,74 +1,56 @@
+<img src="/img/env-vars/variable_list.png" width="100%">
 
-Here you can modify or delete any variable. Changes are saved automatically, so you just have to edit the value.
+__Note:__ Deleting a fleet service variable will not delete a device variable of the same name.
```
`device variable` -> `device service variable`

CameronDiver commented on pages/management/env-vars.md:
```diff
@@ -1,74 +1,56 @@
+## Device Service Variables
 
-Finally, if there are any device-specific variables that **are not** application-wide variables' redefines (i.e. don't have an application-wide variable of the same name) they will be shown at the bottom of the page.
+Device service variables are available to the service container running on that particular device. If both the fleet and the device have a service variable of the same name, the code on the device will use the value of the device service variable. In other words, device service variables redefine (or override) fleet service variables.
```
`available to the service container running` -> `available to serving running`

CameronDiver commented on pages/management/env-vars.md:
```diff
@@ -1,74 +1,56 @@
+<img src="/img/env-vars/device_variables.png" width="100%">
 
-## Managing from the Device Page
+You can override the value of a fleet variable by clicking *override* in the far-right column. This will pop up the variable editing window, where you can change the value:
```
`window` -> `dialog`

CameronDiver commented on pages/runtime/supervisor-api.md:
```diff
@@ -480,6 +480,128 @@ Remotely via the API proxy:
+be able to connnect to the proxy, based on how [redsocks.conf](https://github.com/darkk/redsocks/blob/master/redsocks.conf.example) works. `type` can be `socks4`, `socks5`, `http-connect` or `http-relay` (not all proxies are
+guaranteed to work, especially if they block connections that the resin services may require).
+
+Keep in mind that, even if transparent proxy redirection will take effect immediately after the API call (i.e. all new connections will go through the proxy), open connections will not be closed. So, if for example, the device has managed to connect to the resin VPN without the proxy, it will stay connected directly without trying to reconnect through the proxy, unless the connection breaks - any ...
```
The last sentence here reads kind of weirdly.

CameronDiver commented on pages/runtime/update-locking.md:
```diff
@@ -1,8 +1,8 @@
 ## Update locking
 
-Locking updates means that the Supervisor will not be able to kill your application. This is meant to be used at critical sections of your code where you don't want to be interrupted, or to control that updates are only installed at certain times.
+Locking updates means that the Supervisor will not be able to kill your application. This is meant to be used at critical sections of your code where you don't want to be interrupted, or to ensure that updates are only installed at certain times.
```
I think this should be `will not kill your application` because it can do it, and it will if a lock override is set.

CameronDiver commented on pages/management/env-vars.md:
```diff
@@ -1,74 +1,56 @@
+__Note:__ Deleting a fleet service variable will not delete a device variable of the same name.
 
-Each variable row in the list has an expandable region where you see per-device redefines of this variable, can edit, reset them (the Reset button deletes the redefining per-device variable which automatically makes the application-wide variable seen by the device), or add a new redefine by choosing one of the remaining devices from the drop-down selector.
+If you have already defined service variables at the device level, they will appear in a list below your fleet service variables. You can easily apply a device value to the entire fleet by clicking *Define app-wide*:
```
`Define app-wide` -> `Define fleet-wide`

(Note that this hasn't yet been changed on multistaging, but it will)

zwalchuk commented:
@CameronDiver I went ahead and included the multicontainer changes with some other restructure work I've been working on, so I've addressed your comments but in PR #732
